alf/utils/math_ops.py

```diff
@@ -81,49 +82,54 @@ def swish(x):
 
 
 def max_n(inputs):
-    """Calculate the maximum of n Tensors
+    """Calculate the maximum of n tensors.
 
     Args:
-        inputs (list[Tensor]): list of Tensors, should have the same shape
+        inputs (iterable[Tensor]): an iterable of tensors. It requires that
+            all tensor shapes can be broadcast to the same shape.
     Returns:
-        the elementwise maximum of all the tensors in `inputs`
+        Tensor: the element-wise maximum of all the tensors in ``inputs``.
     """
-    ret = inputs[0]
-    inputs = inputs[1:]
-    for x in inputs:
-        ret = torch.max(ret, x)
-    return ret
+    return functools.reduce(torch.max, inputs, inputs[0])
 
 
 def min_n(inputs):
-    """Calculate the minimum of n Tensors
+    """Calculate the minimum of n tensors.
 
     Args:
-        inputs (list[Tensor]): list of Tensors, should have the same shape
+        inputs (iterable[Tensor]): an iterable of tensors. It requires that
+            all tensor shapes can be broadcast to the same shape.
     Returns:
-        the elementwise minimum of all the tensors in `inputs`
+        Tensor: the element-wise minimum of all the tensors in ``inputs``.
     """
-    ret = inputs[0]
-    inputs = inputs[1:]
-    for x in inputs:
-        ret = torch.min(ret, x)
-    return ret
+    return functools.reduce(torch.min, inputs, inputs[0])
 
 
 def add_n(inputs):
-    """Calculate the sum of n Tensors
+    """Calculate the sum of n tensors.
 
     Args:
-        inputs (list[Tensor]): list of Tensors, should have the same shape
+        inputs (iterable[Tensor]): an iterable of tensors. It requires that
+            all tensor shapes can be broadcast to the same shape.
     Returns:
-        the elementwise sum of all the tensors in `inputs`
+        Tensor: the element-wise sum of all the tensors in ``inputs``.
     """
-    assert isinstance(inputs, list)
-    ret = inputs[0]
-    inputs = inputs[1:]
-    for x in inputs:
-        ret = torch.add(ret, x)
-    return ret
+    return functools.reduce(torch.add, inputs, torch.zeros_like(inputs[0]))
+
+
+def mul_n(inputs):
+    """Calculate the product of n tensors.
+
+    The current solution uses ``functools.reduce``, which is faster than
+    the alternative ``torch.prod(torch.stack(inputs, dim=0), dim=0)``.
+
+    Args:
+        inputs (iterable[Tensor]): an iterable of tensors. It requires that
+            all tensor shapes can be broadcast to the same shape.
+    Returns:
+        Tensor: the element-wise multiplication of all the tensors in ``inputs``.
+    """
+    return functools.reduce(torch.mul, inputs, torch.ones_like(inputs[0]))
```

Comment:
functools.reduce(torch.mul, inputs)

Reply:
No need to use the initilizer. `functools.reduce(torch.mul, inputs)` is better.

Reply:
already updated